new ui

marcelklehr opened the pull request:
![screenshot from 2018-01-28 14-57-16](https://user-images.githubusercontent.com/986878/35482929-fcb57f0e-043b-11e8-8f10-80f75ce1ff0b.png)

- [x] adding bookmarks (fixes #257)
- [x] bookmark search (fixes #131, fixes #297)
- [x] bulk editing (fixes #157)
   - [x] bulk deletion
   - [x] bulk tagging (fixes #12; fixes #286)
- [x] Import & export
   - [x] Import: use fetch() API if available (fixes #423, fixes #80)
   - [x] Use new upload ui (fixes #255 )
- [x] Responsive design (fixes #430)
- [x] editing bookmarks (fixes #437)
- [x] editing tags
- [x] website images (fixes #292)
  - [x] grid view (fixes #55)
- [x] translatability
- [ ] accessibility
- [x] advanced tag filter (fixes #168)
   - [ ] 'Untagged' menu item (do we really need this when we have bulk editing and no more reloads?)
- [ ] ~~sharing (#5)~~

Installation:

```
git clone #-this-repo-#
git checkout feature/new-ui
composer install
npm install
npm run build
```


*old pr: #402*

marcelklehr committed:
First try of backbone/marionette framework

marcelklehr committed:
Add emptyView

marcelklehr committed:
More structure, replace whole template

marcelklehr committed:
Navigation is working

marcelklehr committed:
Allow filter by tag

marcelklehr committed:
Implement 'add bookmark' functionality

marcelklehr committed:
Preliminary detail view for bookmarks

marcelklehr committed:
Implement detail-view#close

marcelklehr committed:
Move add-bookmarks widget to the sidebar

marcelklehr committed:
List view instead of grid, + bulk select tickboxes

marcelklehr committed:
Go back to  grid, this time with flexbox

marcelklehr committed:
Implement search

marcelklehr committed:
UI: Escape url params

marcelklehr committed:
Add bulkActions and implement bulk delete

marcelklehr committed:
Polish selection styles

marcelklehr committed:
[UI] Bulk actions: Add reset button

marcelklehr committed:
UI: Remove magnifying animation

marcelklehr committed:
UI Polish select boxes

marcelklehr committed:
UI: Add clickable tags to bookmarks in overview

marcelklehr committed:
UI: Implement import/export

splitt3r commented:
Nice :) the address field would look pretty awesome if it would fit the navigation style like:

![add](https://user-images.githubusercontent.com/1869584/34895186-d2d75f80-f7e4-11e7-9fef-4c78514c3a9d.png)

And the settings without the button look like:

![settings](https://user-images.githubusercontent.com/1869584/34895229-04e36d3e-f7e5-11e7-88d0-39d3a0b94b94.png)

marcelklehr committed:
Polishing: Add icons

marcelklehr committed:
Refactor bulkActions handling

marcelklehr committed:
Responsive menu

marcelklehr commented:
Thanks :) Good point on the settings :+1: 

re the "new bookmark" form: I originally intended to make it borderless, but wasn't sure whether this would be confusing to people who maybe wouldn't be able to tell that it's a text input. Happy to hear feedback on this, as well as anything else :)

marcelklehr committed:
Restyle: Settings toggle button

splitt3r commented:
It´s not a borderless input it´s an anchor that works like https://docs.nextcloud.com/server/12/developer_manual/design/navigation.html#edit-entry :) More and more apps are adopting the new style guidelines so the user will get used to it and love it ❤️ i think :D

marcelklehr committed:
UI: hide add-bookmark form behind nice nav entry

marcelklehr commented:
Ah, I see, that's nice indeed. Done :)

marcelklehr committed:
UI: Use official app settings styles

marcelklehr committed:
UI: Add quick menu to bookmark cards

marcelklehr committed:
UI: Use the official nc icon classes

marcelklehr committed:
UI: Use external link icon for bm links

marcelklehr committed:
UI: Add "select all" button to bulk actions

marcelklehr committed:
UI: Turn detail view into a drawer

instead of an all-page view

marcelklehr committed:
Basic editing functionality

marcelklehr committed:
Fix bookmark detail view url access

marcelklehr committed:
Implement tag editing

marcelklehr committed:
Refactor detail view management

remove coupling to router
wire up card.edit to detail view

marcelklehr committed:
Fix: directly load 'all bookmarks' view at index

redirect '' to 'all'

marcelklehr committed:
Implement advanced tag filtering

marcelklehr committed:
Implement tag deletion

add more REST-y routes so that backbone can use its default routes
set up event "pipes" so that bookmark updates trigger reload of tags and vice versa

marcelklehr commented:
Quick question: I have now added a special tags view that allow the user to edit, rename and filter by tags. Do you people think, the "favorite tags" should link there directly, with the clicked tag already selected, or should those go into a simple view without all the tag filtering and mangement goodies?

(cc @jancborchardt @Espina2 )

marcelklehr committed:
Remove debug log output

jancborchardt commented:
@marcelklehr the tags should _all_ be listed in the left navigation, sorted by most used. No need for an additional tagging view.

The tag list works like the navigation usually: Clicking on a tag shows all bookmarks with that tag. Clicking another tag switches to that other tag. No combining for now as that led to confusion in the other version.

Let's get this design change out as soon as possible, and let's keep things as simple as possible. :)

marcelklehr commented:
Mh. I think it's likely that people will have *a lot* of tags, so I'm not sure displaying them all in the sidebar is a good idea. (Which is why we opted for *favorite* tags only, as far as I can remember.) Also, I must agree with arthur, in that the whole point of having tags, to me, seems to be that you're able to combine them. Otherwise they're even less useful than folders. After reading #168, I think the problems with the old implementation of filtering were basically:
 - modes in the sidebar
 - tags disappear if you change the filter ("drilldown feature")
 - issues with resetting the app

The implementation I have right now, does neither of these things. Here's a screenshot, for example:
![screenshot from 2018-01-22 14-01-51](https://user-images.githubusercontent.com/986878/35222121-0e29b110-ff7d-11e7-8a62-ca0e023750b9.png)

jancborchardt commented:
Still, why not have all tags in the sidebar? Normal users have around 5-10 tags, and those easily fit in the navigation. Even 20 or more would work nicely with the sorting by number of bookmarks tagged with it.

Combining could even be possible there easily, with an entry in the 3-dot-menu of a tag for it, along with "Rename" and "Delete".

Remember that we are designing Nextcloud for regular people. Our competition or benchmark is stuff like Pocket, or bookmarking in the browser. 

The main point of having tags is to bring some order to your collection. Combining is very advanced, and definitely not the whole point. :)

pixelipo commented:
Regarding tags, I agree with @jancborchardt - I think they should all be listed in a single sidebar.

Another thing - "Tags" in the left sidebar should not be selectable. It should simply be a title for the subsequent list of tags.

And lastly, unrelated to tags, why are bookmarks square boxes? I see no good reason to not have them in a normal vertical list, just like a file list. It seems like all the content would even fit in a single line. Eyes work much faster on a sorted list than a grid, especially on something so non-visual, like this.

Grid would only make sense if the plan is to fetch previews of the websites.

marcelklehr commented:
Thanks for your feedback, folks! :)

The 'tags' menu item is a relic of the specialized tags view, I introduced. I agree that it would be best to have the tags management all in the sidebar instead of a new view, though.

Another quick question on this: What's your opinion on having all selected tags stick to the top of the list? Might be confusing if the locations change.

The idea was to have previews yes. Then it turned out it's quite hard to realize, so we settled on touch icons and that sort of stuff. I'm not sure how best to display those, though, especially since not every site has them. I agree that without visuals it'd be better to have a list. After the tags management I'm going to work on the visuals and see how that pans out. If it doesn't, going back to a list view seems like the best option indeed.

marcelklehr committed:
Move tag management and filtering to sidebar

marcelklehr commented:
![screenshot from 2018-01-23 23-20-24](https://user-images.githubusercontent.com/986878/35304199-f463a196-0094-11e8-9d66-3863b7f8bd79.png)

This is how it looks now. :)

jancborchardt commented:
> Another quick question on this: What's your opinion on having all selected tags stick to the top of the list? Might be confusing if the locations change.

@marcelklehr yeah, as you already say that would be confusing. :) The tags should be sorted by number of bookmarks tagged with it, and so left of the 3-dot-menu there should also be a counter displaying the number of bookmarks.

@pixelipo yep, as @marcelklehr mentioned the plan is to use icons (as preview fetching is hard). A list would be pretty bland.

jancborchardt commented:
@marcelklehr I will look into style detail improvements. Could you check to add the counter to the tags, and also maybe a first quick implementation of the favicon stuff?
Maybe @jakobsack @christophwurst have some pointers here as they did favicon fetching for the Mail app. :)

marcelklehr commented:
I've found a nice drupal module that handles link preview pretty well (fetches meta tags and the largest images in the html as a fallback): https://www.drupal.org/project/external_link_preview
But of course any experience from the mail app team is helpful! :)
I'll look into adding counters, too.

ChristophWurst commented:
> Maybe @jakobsack @ChristophWurst have some pointers here as they did favicon fetching for the Mail app. :)

https://github.com/nextcloud/mail/pull/469/files 😉

marcelklehr committed:
Implement image previews

marcelklehr committed:
UI: Remove paging when fetching bookmarks

marcelklehr committed:
Some style fixes

- make detail view fixed
- bring back blend mode for bm card background images
- bulk actions: make them sticky

marcelklehr commented:
Is there a way to prevent a composer dependency of an app (in this case guzzle) to conflict with an already installed version of that library?

marcelklehr committed:
Add count to tag list

marcelklehr committed:
Fix padding of 'tags' headline

marcelklehr commented:
Thanks @ChristophWurst that was very helpful! I didn't know about ICache :)

![screenshot from 2018-01-28 14-57-16](https://user-images.githubusercontent.com/986878/35482929-fcb57f0e-043b-11e8-8f10-80f75ce1ff0b.png)

One problem remains: The dependency I used to fetch the metadata depends on a different version of guzzle which conflicts with the one nextcloud has installed by default... :/

marcelklehr committed:
Use a fork of link-preview

backported to use guzzle 5.3

ChristophWurst commented:
> Is there a way to prevent a composer dependency of an app (in this case guzzle) to conflict with an already installed version of that library?

No, there is no way to prevent dependency hell in php. I suggest to use the guzzle from server. What do you need it for? Is the `IClientFactory` interface sufficient?

marcelklehr commented:
A library I wanted to use depends on a different version of guzzle, so I forked and backported the library to use guzzle v5.3 :/

marcelklehr committed:
Reload bookmarks when bookmark is added

marcelklehr committed:
Refactor bulk selection

marcelklehr committed:
Fix tag filter

marcelklehr committed:
Polish bookmark-card hover effect

pixelipo commented:
I'm not sure it's such a good idea to have the entire frontend in a single file: `js/bookmarks.js`

It would be nice to have it split up into more files, perhaps moving views outside as well...

marcelklehr commented:
@pixelipo Dead right. That was merely for development, because I didn't want to deal with build steps and all that stuff :D

ChristophWurst commented on templates/main.php:
```diff
@@ -1,8 +1,12 @@
 <?php
+script('bookmarks', '3rdparty/backbone');
```
FYI this will potentially cause problems (either in your app, server components or other apps that are loaded on the same page) as you're overwriting a global js lib, assuming you're installing a newer version of Backbone than the one loaded by the server.

If you go with a build tool like webpack, you can bundle your libraries into a single js that doesn't conflict with other scripts.

marcelklehr replied:
Very true :) Tackled modularization and bundling just now ;)

jancborchardt commented:
@marcelklehr nice progress! :) Looks much better already now! :tada:

marcelklehr committed:
Bulk tagging

marcelklehr committed:
Tags API: Make $count optional

marcelklehr committed:
Use fetch for Import if available

fixes #423

marcelklehr committed:
Fix tagit styles

marcelklehr committed:
Split UI into multiple es6 modules

marcelklehr committed:
Introduce webpack bundling

+ move templates into separate files

marcelklehr committed:
Upadate server templates

lots of stuff no longer needed due to webpack

marcelklehr committed:
Add l10n to templates

and format them to improve legibility

marcelklehr committed:
Fix webpack build

marcelklehr committed:
Replace tagit with select2

+ Fix use of global app var

marcelklehr committed:
Disable 'Shared' & 'Favorites' views for now

marcelklehr commented:
I think sharing bookmarks is something for a different pr. I've also disabled favorites for now, since there's no API for this, yet (and the objective seems to be to get this out ASAP ;) ).

Other than that, I think it's ready for a first review :)

One remark: I didn't even try to comply with the obligatory use of semicolons in js because I wouldn't have been very successful :see_no_evil: , so I've mostly stuck to my usual coding style and am going to convert the code, once it passes your reviews :D

marcelklehr committed:
Fix tags-selection

marcelklehr committed:
Allow scrolling through sidebar tags

marcelklehr commented:
(I've added instructions above on how to build the code :D)

jancborchardt commented:
Cc @nextcloud/javascript and @nextcloud/designers for review :)

pixelipo commented on js/templates/Settings.html:
```diff
@@ -0,0 +1 @@
+<div id="app-settings-header"><button class="settings-button">Settings</a></div><div id="app-settings-content"><form class="import-form" action="bookmark/import" method="post" target="upload_iframe" enctype="multipart/form-data" encoding="multipart/form-data"><input type="file" class="import" name="bm_import" size="5" /><input type="hidden" name="requesttoken" value="'+oc_requesttoken+'" /><butto ...
```
prettify maybe? Whole template in a single line is not very dev-friendly

marcelklehr replied:
I agree!

pixelipo commented on css/bookmarks.css:
```diff
@@ -1,435 +1,429 @@
-    visibility: hidden;
-    float: right;
-    margin: 8px;
+#content > div {
```
Both CSS files should be prettified to meet NC coding standards.

marcelklehr committed:
Pretty-print templates/Settings.html

marcelklehr committed:
Prettify css

marcelklehr committed:
Format js correctly

marcelklehr committed:
Fix Settings template

fix requesttoken interpolation
allow l10n

marcelklehr committed:
Fix TagsManagementTag templates

marcelklehr commented:
@pixelipo done.

pixelipo commented on css/select2.css:
```diff
@@ -0,0 +1 @@
+
```
I guess you can safely delete this file now?

marcelklehr replied:
uh? that was not my intention, actually :D

pixelipo commented on templates/main.php:
```diff
@@ -1,12 +1,8 @@
-script('bookmarks', '3rdparty/js_tpl');
-style('bookmarks', '3rdparty/jquery.tagit');
-
+style('bookmarks', 'select2');
```
`'select2'` is empty, so not needed, right?

pixelipo commented:
Great work, @marcelklehr - code looks really clean now! I will test it on a real NC installation on Monday (can't make it earlier, sorry). Maybe others can...

marcelklehr committed:
Fix select2.css

marcelklehr committed:
Add ImageService

marcelklehr committed:
Fix bookmarks.css

marcelklehr commented:
Also whoever merges this, should probably squash it, no?

marcelklehr committed:
Fix bulk deletion: Would iterate on mutable collection

marcelklehr committed:
Move meta fetching to bm lib

allows import to use it, too

marcelklehr committed:
Reload bookmarks after import

marcelklehr committed:
Fix Bookmarks#isProperURL

marcelklehr committed:
Fix styles for NCv13

marcelklehr committed:
Fix tags management styles

marcelklehr committed:
Fix tabs in bookmarks.css

pixelipo commented:
@marcelklehr usually it's the author that is responsible for squashing in NC

pixelipo commented on js/templates/BookmarkDetail_default.html:
```diff
@@ -0,0 +1,9 @@
+<div class="close icon-close" title="<%- t('bookmarks', 'Close') %>"></div>
```
Why is this file not named using PascalCase, like all the other html files?

marcelklehr replied:
Ah. The idea was to signify that the two files are different modes of the same view. I'm not particular about that, though. If you like, I can just change that to PascalCase.

pixelipo replied:
@marcelklehr I have no strong opinion on that, if `BookmarkDetail_default` is ok for everyone, then leave it be :)

pixelipo commented on js/templates/BookmarkDetail_editing.html:
```diff
@@ -0,0 +1,14 @@
+<div class="close icon-close" title="<%- t('bookmarks', 'Close') %>"></div>
```
same as above

9662 commented:
@jancborchardt 

> The main point of having tags is to bring some order to your collection. Combining is very advanced, and definitely not the whole point. 

FWIW, see this [example story](https://github.com/nextcloud/bookmarks/issues/5#issuecomment-366564046). I would reasonably expect to be able to filter by something like `project-b` + `materials` for example.

9662 commented:
@marcelklehr 

> The idea was to have previews yes. Then it turned out it's quite hard to realize, so we settled on touch icons and that sort of stuff. I'm not sure how best to display those, though, especially since not every site has them.

You could try:
 * Scan bookmark target for JSON-LD, OpenGraph metadata to see if there is an [image](http://schema.org/image) or [logo](http://schema.org/logo) property and fetch that.
 * Scan bookmark target for a `<link rel="icon" …>` element and use that.
 * Blindly go for `/favicon.ico` (does anyone still use that?).
 * Show a generic image.

Personally, I'm just fine with having a vanilla list with one bookmark per row though.

marcelklehr commented:
@9662 Thanks for your feedback.We're indeed using a library to extract the metadata and image from the URLs now :) In the current version of this pr, it's possible to select multiple tags to filter by, as can be seen in https://github.com/nextcloud/bookmarks/pull/424#issuecomment-359953112 ;)

marcelklehr committed:
Fix app.php

jancborchardt commented:
@marcelklehr since it's such a big rewrite it's not necessary to squash. It will help later in case there are some bugs and make it easier to find what caused it. :)

@nextcloud/javascript @blizzz so is this good to go? :)

marcelklehr committed:
Encapsule LinkPreview lib in LinkExplorer service

marcelklehr committed:
Install composer deps on travis

marcelklehr committed:
Fix tests

marcelklehr committed:
Tests: Account for http->https redirects

marcelklehr committed:
Don't store xml errors in memory

marcelklehr committed:
Disable internal errors in LinkExplorer

marcelklehr committed:
Fix tests: www.9gag.com -> 9gag.com

marcelklehr committed:
Fix clickcount API endpoint

marcelklehr commented:
The only test that is still failing is this one, but it doesn't really make sense to me in the first place: https://github.com/nextcloud/bookmarks/blob/dc86ef0bc21ce15388055e4abbc17a05f12f1b39/tests/bookmarkcontroller_test.php#L200 (I know I was one of the last people that touched this part, but I didn't check whether it made sense back then. Shame on me :D)

(Why doesn't it make sense? `%` is an escape character. Registering a url with a non-escaped escape characer and then expecting it to be recognized when that escape character *is* escaped, shouldn't work as those are not the same URLs, IMHO)

marcelklehr committed:
Wire up click counter: Count and sort by clicks

also: open links in new tab

marcelklehr committed:
Sort by lastmodified

sunjam commented:
Well it be possible to view the bookmarks as a compact list ala _Show all Bookmarks_ in Firefox?
    
![80656843-0361-401b-a286-29071100be88](https://user-images.githubusercontent.com/1787238/36633112-6abe50f8-1944-11e8-98d7-5fa17c33912f.png)

marcelklehr committed:
Fix multi-tag filter functionality

jancborchardt commented:
@sunjam no. We are moving forward with this new design, and maintaining two different views is too much. Besides, the compact list is not well-designed:
- takes a lot of width, so not really compact at all.
- title and link are very far away from each other.
- row height is very low so it's not easily clickable.

jancborchardt commented:
@marcelklehr how about we merge this and fix things going forward? It's not a release anyway. But if we keep counting this up to 100 commits, it will never get in. ;)

marcelklehr commented:
@jancborchardt heh, personally I'm happy to merge it, however, I'm not sure anybody besides me has tested the code yet and also, usually @blizzz gets to leave his input about stuff I submit. :)

pixelipo commented:
1. @marcelklehr why is the checkbox inside the 3-dot menu? It's already available on the item itself - no need to duplicate:
![image](https://user-images.githubusercontent.com/4129927/36724128-8109ee06-1bb2-11e8-8619-a03127d142f3.png)

2. Icon should be white:
![image](https://user-images.githubusercontent.com/4129927/36724174-a75fc404-1bb2-11e8-9887-90e7e4126fb4.png)

3. `textarea` should have `resize:vertical` property (disable horizontal resizing):
![image](https://user-images.githubusercontent.com/4129927/36724321-26b92696-1bb3-11e8-8912-d6a9dd70b09e.png)

4. Is there any need for this view in the sidebar? I think that the "Edit" view should be the only view. Ideally, editing would be "inline" and autosave should also exist, but I understand if that is out of scope for this PR.
![image](https://user-images.githubusercontent.com/4129927/36724629-0e10f8ac-1bb4-11e8-9004-40b68b65751e.png)

5. Settings menu looks, hmmm, strange:
![image](https://user-images.githubusercontent.com/4129927/36724846-b8e15574-1bb4-11e8-9bf8-341e59f1b7e8.png)

6. Question for @jancborchardt @skjnldsv : button in sidebar - what is the design guideline for this? It's not defined in the [guidelines](https://docs.nextcloud.com/server/13/developer_manual/design/navigation.html) yet, and we are currently seeing a lot of different solutions there...
![image](https://user-images.githubusercontent.com/4129927/36725050-5ade9fee-1bb5-11e8-9be0-af47d3c931b7.png)
![image](https://user-images.githubusercontent.com/4129927/36725090-79b99766-1bb5-11e8-925b-7580278308c6.png)

7. Few issues regarding the multiselect form (great solution, by the way!):
![image](https://user-images.githubusercontent.com/4129927/36725263-0867084a-1bb6-11e8-89c9-901522321c6b.png)
- lightbue background color does not really fit the design - epsecially when you look at the white arrow pointing to the Bookmarks app icon
- select area is not using default styling (for example, border becomes black on focus)
- "x" button on the right is too small; actually I think it should be removed completely, because it's a very fringe use-case (if you want to rapidly remove a ton tags) and very destructive in nature) CC @jancborchardt 
- "x" button on the tags is too small (see Deck app for comparison)
- "Select all.." icon is too close to the button text
- clicking on a bookmark while the multiselect is visible causes this:
![image](https://user-images.githubusercontent.com/4129927/36725549-ec2f0168-1bb6-11e8-9745-3516e10ebc56.png)
- it looks really messy in mobile view:
![image](https://user-images.githubusercontent.com/4129927/36725688-62b2a9a2-1bb7-11e8-945b-07bd2e60d9bc.png)

8. This menu looks weird. 
![image](https://user-images.githubusercontent.com/4129927/36726223-58fc5320-1bb9-11e8-8deb-52434060316a.png)
In fact, I think that checkbox should be moved out of the 3-dot and in the usual position of the icon (left of text), but we haven't discussed this use-case in @nextcloud/designers Something like this?
![image](https://user-images.githubusercontent.com/4129927/36726167-2447cba0-1bb9-11e8-88ce-9e197e847fcd.png)

***

I think some of the issues I found are serious enough that if the app is deployed as-is, I would call it "alpha"

pixelipo commented:
off topic - I'm unable to test images, because they are not loaded for some reason. @marcelklehr do you have any idea why? Links definitely look weird:
![image](https://user-images.githubusercontent.com/4129927/36726308-94a2aea6-1bb9-11e8-8f35-2ed00aa0dd17.png)

skjnldsv commented:
> Question for @jancborchardt @skjnldsv : button in sidebar - what is the design guideline for this? It's not defined in the guidelines yet, and we are currently seeing a lot of different solutions there...

Guidelines: https://github.com/nextcloud/documentation/pull/684
But it's waiting to be merged here: https://github.com/nextcloud/server/pull/7600

pixelipo commented:
@skjnldsv in the nextcloud/server#7600 "Add group" button is different - is that by design or planned for upgrade?

skjnldsv commented:
@pixelipo it was by design. Last time it was discussed, the outcome was that only one button is supposed to be displayed. :)

Not sure if we want to switch to a secondary button for new groups? :thinking: 
It's not THAT bad tough :)
![capture d ecran_2018-02-27_13-14-56](https://user-images.githubusercontent.com/14975046/36728184-38b125d0-1bc0-11e8-9382-922ec52bb693.png)

MariusBluem commented:
I thought we could kill the Add group button at all ... You can create groups by assigning a non-existing group name to a user already ;) 2 action buttons in the sidebar are to noisy I think.

skjnldsv commented:
I agree @MariusBluem :)

marcelklehr committed:
Fix some minor styling issues

marcelklehr commented:
Thanks for taking the time, Marin :)

1. I think the objective was to allow access to selection to users on touch devices, who don't have hover, as the checkbox on the card is display on hover only.
2. fixed.
3. fixed.
4. The point of the sidebar, besides editing is the ability to view the description, which might turn into a read-it-later-style functionality some day :) I agree that it's not ideal, though. Autosave is not hard to do.
5. fixed :D 
6. I'm not particular about this :) Originally I had a button, then someone suggested the solution as it is now.
7. I agree on all your points here. I'll change the background and fix the styling and the "x"es. That sidebar / bulk editing clash looks bad, indeed. Do you think it would make sense to wire the click on a bookmark card to make it selected without opening the sidebar?
8. Fixed the padding, although your suggestion also looks nice.

marcelklehr commented:
Re: images: The URLs look ok to me. They are hitting an API endpoint that returns the image for a bookmark id. If you're getting 404s on those URLs, that means there was no image found for that site. Maybe try something like a youtube video -- that works for me at least.

marcelklehr committed:
Fix select2 styles

marcelklehr committed:
Fix styling of bulk actions

marcelklehr committed:
TagSelection: Remove clear button

marcelklehr committed:
BookmarkDetailView: In-place editing + auto-save

sunjam commented:
Hi, my instance is currently offline due to disk corruption, so I cannot test the latest build, but I was able to make your request for testing this pull request the top pinned issue on [help.nextcloud.com](https://help.nextcloud.com), so you may soon receive a number of new testers.  Thanks for your great work on the app!

pixelipo commented:
Great work, @marcelklehr !

1. That makes sense - we shouldn't hide actions behind `:hover`. So why not just always show
4. Updated sidebar looks really nice and modern. I do have a couple of issues with it:
  * Delete button doesn't work for me (the one in the 3-dot menu works, though); Maybe it's not even needed? Do we need "Delete" in two different places?
  * "description" doesn't have a visible "edit" visual cue. My suggestion:
```
.description:empty::before {
    content: "Add a description";
    opacity: .3;
}
.description {
    cursor: text;
}
```
This is not translatable, though. String should be injected otherwise.
6. I'm not particular about that button, either. You can keep it for now, or update it, as you wish
7. If you mean "wire it to select only when multiselect popover is visible", then yes, that might be a good idea. It will cause Description not to be accessible during selecting, but that's a fair tradeoff, I think.
8. I would vote for moving the checkbox out of the 3-dot menu and to the left of the tag name. It will half the number of clicks and cursor trips, with no real downside...

Also: Delete button doesn't have a confirmation dialog nor an "Undo" option. It would be worth considering adding one of those (second one being better). Maybe in a future PR?
Autosave feature should have a visual cue that data is saved (check the Deck app for a similar functionality)

guddl commented:
If you have many TAGs -  there should be a scrollbar at the sidebar. On my big screen not all TAGs are shown :grin:

marcelklehr committed:
Add visual cue for editability of description

marcelklehr committed:
Remove menu from  bookmark cards

and always display checkbox

marcelklehr committed:
If selection is active: (Un)select instead of open

marcelklehr committed:
Make tag list scrollable

marcelklehr commented:
@guddl Thanks for the feedback! It's fixed now. :)

marcelklehr committed:
Add saving/saved status to detail view

marcelklehr committed:
Assign background colors to bookmarks without images

marcelklehr committed:
Implement infinite scroll

+ use backbone methods for parsing responses instead of
monkey patching Backbone.sync

marcelklehr committed:
Display spinner if the user has to wait

marcelklehr committed:
Close Details view if user selects a bookmark

marcelklehr committed:
Some style polishing

marcelklehr commented:
@pixelipo Did you get the bookmark images to work? (I also added some URL-based colors for the ones without images, btw)

marcelklehr committed:
Fix export

urlencode(requesttoken)

marcelklehr committed:
Make bookmarklet work

marcelklehr committed:
Styling: Make default card colors more pale

marcelklehr committed:
Remove faulty failing test

background: If the URL contains %ba it is inherently not the same as the same URL with %25ba
thus I think it doesn't make sense to test if they are treated the same

marcelklehr committed:
Make search more useful: interpret blank as AND

marcelklehr commented:
Thanks for your feedback folks! :)

jancborchardt commented:
@marcelklehr yaaaayyyy, finally! :) Congrats!

jospoortvliet commented:
Ok, we should do something with this, blog with some screenshots for example!

marcelklehr commented:
@jospoortvliet Good idea! :) It's not released yet, though, so I'm not sure if it'll be that valuable for the audience. I think a better time for a blog article would probably be the next release, when people can actually install and try the new experience. :)

jospoortvliet commented:
ok, get me some basic info on what changed a few days before the next release! You can write a nice blog if you like but you don't have to - a bullet list of changes is enough, we can write something nice! In this case, links to PR's is actually good enough - and this issue seems an overview so perhaps this will suffice, if it has everything that's new.

Then a few screenshots and we're done :rocket:

bunducafe commented:
Sorry for a kind of off topic question: I activated the bookmark plugin via the nextcloud admin interface. But I cannot see any of the new UI features.
What I get is on the left side just a field for adding an address and the search box for the tags. No bulk deletion, no "all bookmarks", nothing is shown... 

how can I solve this?

marcelklehr commented:
@bunducafe  I'm sorry to tell you that the latest features have not been released yet, but merely staged for inclusion in the next release. The README does indeed not correctly reflect this. Currently, you can only get the new features by manually installing the app (as described in the README).
Don't be afraid to open a new issue for any questions you might have :)

@jospoortvliet Thanks! I will get back to you! :)

e-alfred commented:
@marcelklehr A dead link check would be a great feature I think. Maybe links that throw an error are marked by a tag for example. It could be either done periodically or just with a button to start the check manually.
